Add device_of overload for optional<Tensor>

wenleix opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #54263 Migrate kernels with Tensor? to C10 full dispatcher
* **#54262 Add device_of overload for optional<Tensor>**

register_dispatch_key.py might generate device_of call over
optional<Tensor> if it happened to be the first Tensor-like
argument.

Differential Revision: [D27164093](https://our.internmc.facebook.com/intern/diff/D27164093/)

Commit:
[WIP] Minor fixes to dispatch key registration codegen

1. Handle wrapper methods delegating to legacy kernel when tensor parameter is optional.
2. Handle device guard when tensor parameter is optional.

Differential Revision: [D27164093](https://our.internmc.facebook.com/intern/diff/D27164093/)

[ghstack-poisoned]

Commit:
minor fixes on "[WIP] Minor fixes to dispatch key registration codegen"


1. Handle wrapper methods delegating to legacy kernel when tensor parameter is optional.
2. Handle device guard when tensor parameter is optional.

Differential Revision: [D27164093](https://our.internmc.facebook.com/intern/diff/D27164093/)

[ghstack-poisoned]

wenleix commented on tools/codegen/dest/register_dispatch_key.py:
```diff
@@ -121,7 +121,14 @@ def generate_defn(cpp_sig: CppSignature) -> str:
+            args_exprs = []
+            for a in args:
+                # TODO: this is a big HACK, figure out better way...
+                if 'legacy' in impl_name and 'optional<Tensor>' in a.type:
```
cc @smessmer  about current hack.

Commit:
squash & rebase on "[WIP] Minor fixes to dispatch key registration codegen"


1. Handle wrapper methods delegating to legacy kernel when tensor parameter is optional.
2. Handle device guard when tensor parameter is optional.

Differential Revision: [D27164093](https://our.internmc.facebook.com/intern/diff/D27164093/)

[ghstack-poisoned]

wenleix commented on tools/codegen/dest/register_dispatch_key.py:
```diff
@@ -161,9 +166,13 @@ def generate_defn(cpp_sig: CppSignature) -> str:
     {cuda_guard_from_tensor_options}
 """
-                elif f.device_guard and device_of is not None:
+                elif f.device_guard and tensor_arg_name is not None:
```
another approach is to still use the first `tesnor_like` parameter, and use `device_of` or `device_or_default` based on whether it's optional.

wenleix replied:
An alternative approach would be still use the first tensor-like argument, but generate `device_of` or `device_or_default` depends on whether it's optional or not.

Commit:
rebase & add note on "Minor fix to dispatch key registration codegen"


Handle device guard when tensor parameter is optional.

Differential Revision: [D27164093](https://our.internmc.facebook.com/intern/diff/D27164093/)

[ghstack-poisoned]

ezyang commented:
Can you say which schemas are affected by this?

ezyang commented:
The adjustment to codegen seems overwrought here. Why not just add a `device_of(optional<Tensor>)` overload to ATen/DeviceGuard.h?

wenleix commented:
@ezyang : 

> Can you say which schemas are affected by this?

`_thnn_differentiable_lstm_cell_backward`. Here is generated `RegisterMath.cpp`:

```c++
std::tuple<Tensor,Tensor,Tensor,Tensor,Tensor> wrapper__thnn_differentiable_lstm_cell_backward(const c10::optional<Tensor> & grad_hy, const c10::optional<Tensor> & grad_cy, const Tensor & input_gates, const Tensor & hidden_gates, const c10::optional<Tensor> & input_bias, const c10::optional<Tensor> & hidden_bias, const Tensor & cx, const Tensor & cy) {
    const OptionalDeviceGuard device_guard(device_of(grad_hy));
    return at::native::_thnn_differentiable_lstm_cell_backward(grad_hy, grad_cy, input_gates, hidden_gates, input_bias, hidden_bias, cx, cy);
}
```

`device_of(grad_hy)` complaints, as `grad_hy` is now a `Optional<Tensor>`

Commit:
rebase && rework first diff based on comment on "Minor fix to dispatch key registration codegen"


Handle device guard when tensor parameter is optional.

Differential Revision: [D27164093](https://our.internmc.facebook.com/intern/diff/D27164093/)

[ghstack-poisoned]

ezyang commented:
I don't think you should fix this in this PR, but there is probably a latent bug in the generated device guards here. Suppose you call this function with nullopt in first argument. Then it will always skip setting the device guard. But if there are other arguments, we ought to have set the device guard. So really what you need to do is find a non-optional tensor argument, or failing that, find the first non-nullopt tensor argument to test. Don't fix this in this PR.
